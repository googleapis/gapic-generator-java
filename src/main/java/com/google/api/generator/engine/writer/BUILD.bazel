package(default_visibility = ["//visibility:public"])

filegroup(
    name = "writer_files",
    srcs = glob(["*.java"]),
)

java_library(
    name = "writer",
    srcs = [
        ":writer_files",
    ],
    deps = [
        "//src/main/java/com/google/api/generator/engine/ast",
        "//src/main/java/com/google/api/generator/engine/lexicon",
        "@com_google_guava_guava//jar",
        "@google_java_format_all_deps//jar",
    ],
)
