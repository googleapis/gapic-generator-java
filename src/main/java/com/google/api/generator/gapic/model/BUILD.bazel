load("@rules_java//java:defs.bzl", "java_library")

package(default_visibility = ["//visibility:public"])

filegroup(
    name = "model_files",
    srcs = glob(["*.java"]),
)

java_library(
    name = "model",
    srcs = [
        ":model_files",
    ],
    deps = [
        "//:service_config_java_proto",
        "//src/main/java/com/google/api/generator:autovalue",
        "//src/main/java/com/google/api/generator/engine/ast",
        "//src/main/java/com/google/api/generator/gapic/utils",
        "@com_google_api_api_common//jar",
        "@com_google_auto_value_auto_value//jar",
        "@com_google_auto_value_auto_value_annotations//jar",
        "@com_google_code_findbugs_jsr305//jar",
        "@com_google_googleapis//google/api:api_java_proto",
        "@com_google_googleapis//google/rpc:rpc_java_proto",
        "@com_google_guava_guava//jar",
        "@com_google_protobuf//:protobuf_java",
    ],
)
