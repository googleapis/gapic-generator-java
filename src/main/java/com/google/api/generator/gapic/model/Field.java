// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.api.generator.engine.ast.TypeNode;
import com.google.auto.value.AutoValue;
import javax.annotation.Nullable;

@AutoValue
public abstract class Field {
  public abstract String name();

  public abstract TypeNode type();

  public abstract boolean isRepeated();

  @Nullable
  public abstract ResourceReference resourceReference();

  public boolean hasResourceReference() {
    return type().equals(TypeNode.STRING) && resourceReference() != null;
  }

  public static Builder builder() {
    return new AutoValue_Field.Builder().setIsRepeated(false);
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setName(String name);

    public abstract Builder setType(TypeNode type);

    public abstract Builder setIsRepeated(boolean isRepeated);

    public abstract Builder setResourceReference(ResourceReference resourceReference);

    public abstract Field build();
  }
}
