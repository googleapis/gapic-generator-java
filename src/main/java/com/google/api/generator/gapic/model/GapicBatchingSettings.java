// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.auto.value.AutoValue;
import javax.annotation.Nullable;

@AutoValue
public abstract class GapicBatchingSettings {
  public enum FlowControlLimitExceededBehavior {
    THROW_EXCEPTION,
    BLOCK,
    IGNORE
  };

  public abstract String protoPakkage();

  public abstract String serviceName();

  public abstract String methodName();

  public abstract int elementCountThreshold();

  public abstract long requestByteThreshold();

  public abstract long delayThresholdMillis();

  @Nullable
  public abstract Integer flowControlElementLimit();

  @Nullable
  public abstract Integer flowControlByteLimit();

  public abstract FlowControlLimitExceededBehavior flowControlLimitExceededBehavior();

  public static Builder builder() {
    return new AutoValue_GapicBatchingSettings.Builder()
        .setFlowControlLimitExceededBehavior(FlowControlLimitExceededBehavior.IGNORE);
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setProtoPakkage(String protoPakkage);

    public abstract Builder setServiceName(String serviceName);

    public abstract Builder setMethodName(String methodName);

    public abstract Builder setElementCountThreshold(int elementCountThreshold);

    public abstract Builder setRequestByteThreshold(long requestByteThreshold);

    public abstract Builder setDelayThresholdMillis(long delalyThresholdMillis);

    public abstract Builder setFlowControlElementLimit(Integer flowControlElementLimit);

    public abstract Builder setFlowControlByteLimit(Integer flowControlByteLimit);

    public abstract Builder setFlowControlLimitExceededBehavior(
        FlowControlLimitExceededBehavior behavior);

    public abstract GapicBatchingSettings build();
  }
}
