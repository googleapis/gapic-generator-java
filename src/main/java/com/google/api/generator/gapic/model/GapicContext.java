// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.List;
import java.util.Map;

@AutoValue
public abstract class GapicContext {
  // Maps the message name (as it appears in the protobuf) to Messages.
  public abstract ImmutableMap<String, Message> messages();

  // Maps the resource type string to ResourceNames.
  public abstract ImmutableMap<String, ResourceName> resourceNames();

  public abstract ImmutableList<Service> services();

  public static Builder builder() {
    return new AutoValue_GapicContext.Builder();
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setMessages(Map<String, Message> messages);

    public abstract Builder setResourceNames(Map<String, ResourceName> resourceNames);

    public abstract Builder setServices(List<Service> services);

    public abstract GapicContext build();
  }
}
