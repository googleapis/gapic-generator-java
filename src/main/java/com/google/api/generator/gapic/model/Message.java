// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.api.generator.engine.ast.TypeNode;
import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@AutoValue
public abstract class Message {
  public abstract String name();

  // TODO(unsupported): oneof fields are parsed as separate ones because field flattening refers to
  // a specific field.
  public abstract ImmutableList<Field> fields();

  public abstract TypeNode type();

  public abstract ImmutableMap<String, Field> fieldMap();

  public static Builder builder() {
    return new AutoValue_Message.Builder();
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setName(String name);

    public abstract Builder setFields(List<Field> fields);

    public abstract Builder setType(TypeNode type);

    abstract Builder setFieldMap(Map<String, Field> fieldMap);

    abstract ImmutableList<Field> fields();

    abstract Message autoBuild();

    public Message build() {
      setFieldMap(fields().stream().collect(Collectors.toMap(f -> f.name(), f -> f)));
      return autoBuild();
    }
  }
}
