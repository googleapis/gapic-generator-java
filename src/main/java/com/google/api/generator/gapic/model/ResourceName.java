// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.model;

import com.google.api.generator.engine.ast.TypeNode;
import com.google.api.generator.engine.ast.VaporReference;
import com.google.api.generator.gapic.utils.JavaStyle;
import com.google.api.generator.gapic.utils.ResourceNameConstants;
import com.google.auto.value.AutoValue;
import com.google.common.collect.ImmutableList;
import java.util.List;
import java.util.Objects;
import javax.annotation.Nullable;

@AutoValue
public abstract class ResourceName {
  static final String SLASH = "/";

  // The original binding variable name.
  // This should be in lower_snake_case in the proto, and expected to be surrounded by braces.
  // Example: In projects/{project}/billingAccounts/billing_account, the variable name would be
  // "billing_account."
  public abstract String variableName();

  // The Java package of the project, or that of a subpackage where the resource name was defined.
  // That is, resource names defined outside of this project will still have the project's package.
  public abstract String pakkage();

  // The resource type.
  public abstract String resourceTypeString();

  // A list of patterns such as projects/{project}/locations/{location}/resources/{this_resource}.
  public abstract ImmutableList<String> patterns();

  // The Java TypeNode of the resource name helper class to generate.
  public abstract TypeNode type();

  public abstract boolean isOnlyWildcard();

  // The message in which this resource was defined. Optional.
  // This is expected to be empty for file-level definitions.
  @Nullable
  public abstract String parentMessageName();

  public boolean hasParentMessageName() {
    return parentMessageName() != null;
  }

  public static Builder builder() {
    return new AutoValue_ResourceName.Builder().setIsOnlyWildcard(false);
  }

  public static ResourceName createWildcard(String resourceTypeString, String pakkage) {
    String placeholderVarName =
        JavaStyle.toLowerCamelCase(
            resourceTypeString.substring(resourceTypeString.indexOf(SLASH) + 1));
    return builder()
        .setVariableName(placeholderVarName)
        .setPakkage(pakkage)
        .setResourceTypeString(resourceTypeString)
        .setPatterns(ImmutableList.of(ResourceNameConstants.WILDCARD_PATTERN))
        .setIsOnlyWildcard(true)
        .setType(
            TypeNode.withReference(
                VaporReference.builder()
                    .setName("ResourceName")
                    .setPakkage("com.google.api.resourcenames")
                    .build()))
        .build();
  }

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof ResourceName)) {
      return false;
    }

    ResourceName other = (ResourceName) o;
    return variableName().equals(other.variableName())
        && pakkage().equals(other.pakkage())
        && resourceTypeString().equals(other.resourceTypeString())
        && patterns().equals(other.patterns())
        && Objects.equals(parentMessageName(), other.parentMessageName())
        && Objects.equals(type(), other.type());
  }

  @Override
  public int hashCode() {
    int parentMessageNameHashCode =
        parentMessageName() == null ? 0 : parentMessageName().hashCode();
    int typeHashCode = type() == null ? 0 : type().hashCode();
    return 17 * variableName().hashCode()
        + 19 * pakkage().hashCode()
        + 23 * resourceTypeString().hashCode()
        + 31 * patterns().hashCode()
        + 37 * parentMessageNameHashCode
        + 41 * typeHashCode;
  }

  @AutoValue.Builder
  public abstract static class Builder {
    public abstract Builder setVariableName(String variableName);

    public abstract Builder setPakkage(String pakkage);

    public abstract Builder setResourceTypeString(String resourceTypeString);

    public abstract Builder setPatterns(List<String> patterns);

    public abstract Builder setParentMessageName(String parentMessageName);

    // Private setters.
    abstract Builder setType(TypeNode type);

    abstract Builder setIsOnlyWildcard(boolean isOnlyWildcard);

    // Private accessors.
    abstract String pakkage();

    abstract String resourceTypeString();

    abstract boolean isOnlyWildcard();

    // Private.
    abstract ResourceName autoBuild();

    public ResourceName build() {
      if (!isOnlyWildcard()) {
        String typeName =
            resourceTypeString().substring(resourceTypeString().lastIndexOf(SLASH) + 1);
        setType(
            TypeNode.withReference(
                VaporReference.builder()
                    .setName(String.format("%sName", typeName))
                    .setPakkage(pakkage())
                    .build()));
      }
      return autoBuild();
    }
  }
}
