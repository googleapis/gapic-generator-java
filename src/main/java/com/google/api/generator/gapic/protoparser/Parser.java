// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.protoparser;

import com.google.api.generator.engine.ast.TypeNode;
import com.google.api.generator.engine.ast.VaporReference;
import com.google.api.generator.gapic.model.Field;
import com.google.api.generator.gapic.model.Message;
import com.google.api.generator.gapic.model.Method;
import com.google.api.generator.gapic.model.Service;
import com.google.common.base.Preconditions;
import com.google.common.collect.Maps;
import com.google.protobuf.DescriptorProtos.FileDescriptorProto;
import com.google.protobuf.Descriptors.Descriptor;
import com.google.protobuf.Descriptors.DescriptorValidationException;
import com.google.protobuf.Descriptors.FileDescriptor;
import com.google.protobuf.Descriptors.MethodDescriptor;
import com.google.protobuf.Descriptors.ServiceDescriptor;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class Parser {
  static class GapicParserException extends RuntimeException {
    public GapicParserException(String errorMessage) {
      super(errorMessage);
    }
  }

  public static List<Service> parseServices(
      CodeGeneratorRequest request, Map<String, Message> messageTypes) {
    Map<String, FileDescriptor> fileDescriptors = getFilesToGenerate(request);
    List<Service> services = new ArrayList<>();
    for (String fileToGenerate : request.getFileToGenerateList()) {
      FileDescriptor fileDescriptor =
          Preconditions.checkNotNull(
              fileDescriptors.get(fileToGenerate),
              "Missing file descriptor for [%s]",
              fileToGenerate);

      String pakkage = TypeParser.getPackage(fileDescriptor);
      for (ServiceDescriptor serviceDescriptor : fileDescriptor.getServices()) {
        List<Method> methods = parseMethods(serviceDescriptor, messageTypes);
        Service service =
            Service.builder()
                .setName(serviceDescriptor.getName())
                .setPakkage(pakkage)
                .setMethods(methods)
                .build();
        services.add(service);
      }
    }

    return services;
  }

  public static Map<String, Message> parseMessages(CodeGeneratorRequest request) {
    Map<String, FileDescriptor> fileDescriptors = getFilesToGenerate(request);
    Map<String, Message> messages = new HashMap<>();
    for (String fileToGenerate : request.getFileToGenerateList()) {
      FileDescriptor fileDescriptor =
          Preconditions.checkNotNull(
              fileDescriptors.get(fileToGenerate),
              "Missing file descriptor for [%s]",
              fileToGenerate);

      String pakkage = TypeParser.getPackage(fileDescriptor);
      for (Descriptor messageDescriptor : fileDescriptor.getMessageTypes()) {
        List<Field> fields = parseFields(messageDescriptor);
        String messageName = messageDescriptor.getName();
        messages.put(
            messageName,
            Message.builder()
                .setType(
                    TypeNode.withReference(
                        VaporReference.builder().setName(messageName).setPakkage(pakkage).build()))
                .setName(messageName)
                .setFields(fields)
                .build());
      }
    }
    return messages;
  }

  private static List<Field> parseFields(Descriptor messageDescriptor) {
    return messageDescriptor.getFields().stream()
        .map(f -> Field.builder().setName(f.getName()).setType(TypeParser.parseType(f)).build())
        .collect(Collectors.toList());
  }

  private static Map<String, FileDescriptor> getFilesToGenerate(CodeGeneratorRequest request) {
    // Build the fileDescriptors map so that we can create the FDs for the filesToGenerate.
    Map<String, FileDescriptor> fileDescriptors = Maps.newHashMap();
    for (FileDescriptorProto fileDescriptorProto : request.getProtoFileList()) {
      // Look up the imported files from previous file descriptors.  It is sufficient to look at
      // only previous file descriptors because CodeGeneratorRequest guarantees that the files
      // are sorted in topological order.
      FileDescriptor[] deps = new FileDescriptor[fileDescriptorProto.getDependencyCount()];
      for (int i = 0; i < fileDescriptorProto.getDependencyCount(); i++) {
        String name = fileDescriptorProto.getDependency(i);
        deps[i] =
            Preconditions.checkNotNull(
                fileDescriptors.get(name), "Missing file descriptor for [%s]", name);
      }

      FileDescriptor fileDescriptor = null;
      try {
        fileDescriptor = FileDescriptor.buildFrom(fileDescriptorProto, deps);
      } catch (DescriptorValidationException e) {
        throw new GapicParserException(e.getMessage());
      }

      fileDescriptors.put(fileDescriptor.getName(), fileDescriptor);
    }
    return fileDescriptors;
  }

  private static List<Method> parseMethods(
      ServiceDescriptor serviceDescriptor, Map<String, Message> messageTypes) {
    List<Method> methods = new ArrayList<>();
    for (MethodDescriptor methodDescriptor : serviceDescriptor.getMethods()) {
      methods.add(
          Method.builder()
              .setName(methodDescriptor.getName())
              .setInputType(TypeParser.parseType(methodDescriptor.getInputType()))
              .setOutputType(TypeParser.parseType(methodDescriptor.getOutputType()))
              .build());
    }
    return methods;
  }
}
