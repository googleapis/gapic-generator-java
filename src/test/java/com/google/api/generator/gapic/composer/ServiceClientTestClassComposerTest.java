// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.api.generator.gapic.composer;

import static junit.framework.Assert.assertEquals;

import com.google.api.generator.engine.writer.JavaWriterVisitor;
import com.google.api.generator.gapic.model.GapicClass;
import com.google.api.generator.gapic.model.Message;
import com.google.api.generator.gapic.model.ResourceName;
import com.google.api.generator.gapic.model.Service;
import com.google.api.generator.gapic.protoparser.Parser;
import com.google.protobuf.Descriptors.FileDescriptor;
import com.google.protobuf.Descriptors.ServiceDescriptor;
import com.google.showcase.v1beta1.EchoOuterClass;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.junit.Before;
import org.junit.Test;

public class ServiceClientTestClassComposerTest {
  private ServiceDescriptor echoService;
  private FileDescriptor echoFileDescriptor;

  @Before
  public void setUp() {
    echoFileDescriptor = EchoOuterClass.getDescriptor();
    echoService = echoFileDescriptor.getServices().get(0);
    assertEquals(echoService.getName(), "Echo");
  }

  @Test
  public void generateServiceClasses() {
    Map<String, Message> messageTypes = Parser.parseMessages(echoFileDescriptor);
    Map<String, ResourceName> resourceNames = Parser.parseResourceNames(echoFileDescriptor);
    Set<ResourceName> outputResourceNames = new HashSet<>();
    List<Service> services =
        Parser.parseService(echoFileDescriptor, messageTypes, resourceNames, outputResourceNames);

    Service echoProtoService = services.get(0);
    GapicClass clazz =
        ServiceClientTestClassComposer.instance()
            .generate(echoProtoService, resourceNames, messageTypes);

    JavaWriterVisitor visitor = new JavaWriterVisitor();
    clazz.classDefinition().accept(visitor);
    assertEquals(EXPECTED_CLASS_STRING, visitor.write());
  }

  // TODO(miraleung): Update this when a file-diffing test mechanism is in place.
  private static final String EXPECTED_CLASS_STRING =
      "package com.google.showcase.v1beta1;\n"
          + "\n"
          + "import com.google.api.gax.core.NoCredentialsProvider;\n"
          + "import com.google.api.gax.grpc.testing.LocalChannelProvider;\n"
          + "import com.google.api.gax.grpc.testing.MockGrpcService;\n"
          + "import com.google.api.gax.grpc.testing.MockServiceHelper;\n"
          + "import com.google.api.gax.rpc.InvalidArgumentException;\n"
          + "import com.google.api.resourcenames.ResourceName;\n"
          + "import com.google.rpc.Status;\n"
          + "import io.grpc.StatusRuntimeException;\n"
          + "import java.io.IOException;\n"
          + "import java.util.Arrays;\n"
          + "import java.util.UUID;\n"
          + "import javax.annotation.Generated;\n"
          + "import org.junit.After;\n"
          + "import org.junit.AfterClass;\n"
          + "import org.junit.Assert;\n"
          + "import org.junit.Before;\n"
          + "import org.junit.BeforeClass;\n"
          + "import org.junit.Test;\n"
          + "\n"
          + "@Generated(\"by gapic-generator-java\")\n"
          + "public class EchoClientTest {\n"
          + "  public static MockServiceHelper mockServiceHelper;\n"
          + "  public static MockEcho mockEcho;\n"
          + "  public EchoClient client;\n"
          + "  public LocalChannelProvider channelProvider;\n"
          + "\n"
          + "  @BeforeClass\n"
          + "  public static void startStaticServer() {\n"
          + "    mockEcho = new MockEcho();\n"
          + "    mockServiceHelper =\n"
          + "        new MockServiceHelper(\n"
          + "            UUID.randomUUID().toString(),"
          + " Arrays.<MockGrpcService>asList(mockEcho));\n"
          + "    mockServiceHelper.start();\n"
          + "  }\n"
          + "\n"
          + "  @AfterClass\n"
          + "  public static void stopServer() {\n"
          + "    mockServiceHelper.stop();\n"
          + "  }\n"
          + "\n"
          + "  @Before\n"
          + "  public void setUp() throws IOException {\n"
          + "    mockServiceHelper.reset();\n"
          + "    channelProvider = mockServiceHelper.createChannelProvider();\n"
          + "    EchoSettings settings =\n"
          + "        EchoSettings.newBuilder()\n"
          + "            .setTransportChannelProvider(channelProvider)\n"
          + "            .setCredentialsProvider(NoCredentialsProvider.create())\n"
          + "            .build();\n"
          + "    client = EchoClient.create(settings);\n"
          + "  }\n"
          + "\n"
          + "  @After\n"
          + "  public void tearDown() throws Exception {\n"
          + "    client.close();\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void echoExceptionTest() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    mockEcho.addException(exception);\n"
          + "    try {\n"
          + "      String content = \"content951530617\";\n"
          + "      client.echo(content);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void echoExceptionTest2() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    mockEcho.addException(exception);\n"
          + "    try {\n"
          + "      Status error = Status.newBuilder().build();\n"
          + "      client.echo(error);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void echoExceptionTest3() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    mockEcho.addException(exception);\n"
          + "    try {\n"
          + "      String content = \"content951530617\";\n"
          + "      Severity severity = Severity.forNumber(0);\n"
          + "      client.echo(content, severity);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void echoExceptionTest4() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    mockEcho.addException(exception);\n"
          + "    try {\n"
          + "      String name = \"name3373707\";\n"
          + "      client.echo(name);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void echoExceptionTest5() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    mockEcho.addException(exception);\n"
          + "    try {\n"
          + "      FoobarName name = FoobarName.ofProjectFoobarName(\"[PROJECT]\","
          + " \"[FOOBAR]\");\n"
          + "      client.echo(name);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void echoExceptionTest6() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    mockEcho.addException(exception);\n"
          + "    try {\n"
          + "      ResourceName parent = FoobarName.ofProjectFoobarName(\"[PROJECT]\","
          + " \"[FOOBAR]\");\n"
          + "      client.echo(parent);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void echoExceptionTest7() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    mockEcho.addException(exception);\n"
          + "    try {\n"
          + "      String parent = \"parent-995424086\";\n"
          + "      client.echo(parent);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void expandExceptionTest() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    addException(exception);\n"
          + "    try {\n"
          + "      String content = \"content951530617\";\n"
          + "      Status error = Status.newBuilder().build();\n"
          + "      client.expand(content, error);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void collectExceptionTest() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    addException(exception);\n"
          + "    try {\n"
          + "      String content = \"content951530617\";\n"
          + "      client.collect(content);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "\n"
          + "  @Test\n"
          + "  public void chatAgainExceptionTest() throws Exception {\n"
          + "    StatusRuntimeException exception = new"
          + " StatusRuntimeException(io.grpc.Status.INVALID_ARGUMENT);\n"
          + "    addException(exception);\n"
          + "    try {\n"
          + "      String content = \"content951530617\";\n"
          + "      client.chatAgain(content);\n"
          + "      Assert.fail(\"No exception raised\");\n"
          + "    } catch (InvalidArgumentException e) {\n"
          + "      // Expected exception.\n"
          + "    }\n"
          + "  }\n"
          + "}\n";
}
