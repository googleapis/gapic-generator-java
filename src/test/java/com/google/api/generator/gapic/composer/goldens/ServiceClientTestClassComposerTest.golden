package com.google.showcase.v1beta1;

import com.google.api.gax.core.NoCredentialsProvider;
import com.google.api.gax.grpc.testing.LocalChannelProvider;
import com.google.api.gax.grpc.testing.MockGrpcService;
import com.google.api.gax.grpc.testing.MockServiceHelper;
import java.io.IOException;
import java.util.Arrays;
import java.util.UUID;
import javax.annotation.Generated;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;

@Generated("by gapic-generator-java")
public class EchoClientTest {
  public static MockServiceHelper mockServiceHelper;
  public static MockEcho mockEcho;
  public EchoClient client;
  public LocalChannelProvider channelProvider;

  @BeforeClass
  public static void startStaticServer() {
    mockEcho = new MockEcho();
    mockServiceHelper =
        new MockServiceHelper(
            UUID.randomUUID().toString(), Arrays.<MockGrpcService>asList(mockEcho));
    mockServiceHelper.start();
  }

  @AfterClass
  public static void stopServer() {
    mockServiceHelper.stop();
  }

  @Before
  public void setUp() throws IOException {
    mockServiceHelper.reset();
    channelProvider = mockServiceHelper.createChannelProvider();
    EchoSettings settings =
        EchoSettings.newBuilder()
            .setTransportChannelProvider(channelProvider)
            .setCredentialsProvider(NoCredentialsProvider.create())
            .build();
    client = EchoClient.create(settings);
  }

  @After
  public void tearDown() throws Exception {
    client.close();
  }
}
